projects/RabbitMQ.Client/client/impl/OutgoingCommand.cs

```diff
@@ -31,49 +31,75 @@
 
 using System;
 using System.Buffers;
-using RabbitMQ.Client.Framing.Impl;
 
 namespace RabbitMQ.Client.Impl
 {
-    internal readonly struct OutgoingCommand
+    internal interface IOutgoingCommand
     {
-        // EmptyFrameSize, 8 = 1 + 2 + 4 + 1
-        // - 1 byte of frame type
-        // - 2 bytes of channel number
-        // - 4 bytes of frame payload length
-        // - 1 byte of payload trailer FrameEnd byte
-        private const int EmptyFrameSize = 8;
+        MethodBase Method { get; }
+        ReadOnlyMemory<byte> SerializeToFrames(ushort channelNumber, uint frameMax);
+    }
 
-        public readonly MethodBase Method;
-        private readonly ContentHeaderBase _header;
-        private readonly ReadOnlyMemory<byte> _body;
+    internal readonly struct OutgoingCommand : IOutgoingCommand
+    {
+        private readonly MethodBase _method;
+        public MethodBase Method => _method;
 
         public OutgoingCommand(MethodBase method)
-            : this(method, null, ReadOnlyMemory<byte>.Empty)
         {
+            _method = method;
         }
 
-        public OutgoingCommand(MethodBase method, ContentHeaderBase header, ReadOnlyMemory<byte> body)
+        public ReadOnlyMemory<byte> SerializeToFrames(ushort channelNumber, uint frameMax)
         {
-            Method = method;
+            int size = Method.GetRequiredBufferSize() + Framing.Method.FrameSize;
+
+            // Will be returned by SocketFrameWriter.WriteLoop
+            byte[] rentedArray = ArrayPool<byte>.Shared.Rent(size);
+            Span<byte> span = rentedArray.AsSpan(0, size);
```

Comment:
technically we don't need the span, we can just create the span over the whole array, since we don't check the size at the end, we also don't need to limit it.

Reply:
Fixed